File: apps/worker/engine/credentials.py
"""Populate os.environ from systemd LoadCredential= files (reference/systemd,
ADR-130) when running under systemd (CREDENTIALS_DIRECTORY set), so existing
os.environ.get('DATABASE_URL') / os.environ.get('SMTP_PASSWORD') call sites
need no changes. In local/dev, where CREDENTIALS_DIRECTORY is unset, this is
a no-op and the existing dotenv-based loading takes over instead.
"""
import os

CREDENTIAL_ENV_MAP = {
    'database_url': 'DATABASE_URL',
    'smtp_password': 'SMTP_PASSWORD',
}


def load_credentials_into_env():
    cred_dir = os.environ.get('CREDENTIALS_DIRECTORY')
    if not cred_dir:
        return
    for credential_name, env_var in CREDENTIAL_ENV_MAP.items():
        path = os.path.join(cred_dir, credential_name)
        try:
            with open(path, 'r', encoding='utf-8') as f:
                os.environ[env_var] = f.read().strip()
        except OSError as err:
            raise RuntimeError(
                f"Failed to read credential '{credential_name}' from {cred_dir}: {err}"
            ) from err

File: apps/api/src/utils/credentials.js
import fs from 'fs';
import path from 'path';

// Populates process.env from systemd LoadCredential= files (reference/systemd,
// ADR-130) when running under systemd (CREDENTIALS_DIRECTORY set), so
// existing process.env.DATABASE_URL / process.env.JWT_SECRET call sites need
// no changes. In local/dev, where CREDENTIALS_DIRECTORY is unset, this is a
// no-op and the existing .env-based loading takes over instead.
const CREDENTIAL_ENV_MAP = {
	database_url: 'DATABASE_URL',
	jwt_secret: 'JWT_SECRET',
};

export function loadCredentialsIntoEnv() {
	const dir = process.env.CREDENTIALS_DIRECTORY;
	if (!dir) return;
	for (const [credentialName, envVar] of Object.entries(CREDENTIAL_ENV_MAP)) {
		try {
			process.env[envVar] = fs.readFileSync(path.join(dir, credentialName), 'utf8').trim();
		} catch (err) {
			throw new Error(`Failed to read credential '${credentialName}' from ${dir}: ${err.message}`);
		}
	}
}
